cloudpickle/cloudpickle_fast.py

```diff
@@ -0,0 +1,605 @@
+        pickle.dumps(obj)
+    except Exception:
+        if obj.__module__ == "builtins":
+            if obj in _BUILTIN_TYPE_NAMES:
+                return _builtin_type_reduce(obj)
+
+        typ = type(obj)
+        if typ is not obj and isinstance(obj, type):  # noqa: E721
+            return _dynamic_class_reduce(obj)
+
+    else:
+        # if pickle.dumps worked out fine, then simply fallback to the
+        # traditional pickle by attribute # implemented in the builtin
+        # `Pickler.save_global`.
+        return NotImplementedError
+
+
+# COLLECTIONS OF OBJECTS STATE SETTERS
+# ------------------------------------
+# state setters are called at unpickling time, once the object is created and
+# it has to be updated to how it was at unpickling time.
+
+
+def _function_setstate(obj, state):
+    """Update the state of a dynaamic function.
+
+    As __closure__ and __globals__ are readonly attributes of a function, we
+    cannot rely on the native setstate routine of pickle.load_build, that calls
+    setattr on items of the slotstate. Instead, we have to modify them inplace.
+    """
+    state, slotstate = state
+    obj.__dict__.update(state)
+
+    obj_globals = slotstate.pop("__globals__")
+    obj_closure = slotstate.pop("__closure__")
+
+    # remove uncessary  references to submodules
+    obj_globals.pop("__submodules__")
+    obj.__globals__.update(obj_globals)
+    obj.__globals__["__builtins__"] = __builtins__
+
+    if obj_closure is not None:
+        for i, cell in enumerate(obj_closure):
+            try:
+                value = cell.cell_contents
+            except ValueError:  # cell is empty
+                continue
+            obj.__closure__[i].cell_contents = value
+
+    for k, v in slotstate.items():
+        setattr(obj, k, v)
+
+
+def _class_setstate(obj, state):
+    state, slotstate = state
+    registry = None
+    for attrname, attr in state.items():
+        if attrname == "_abc_impl":
+            registry = attr
+        else:
+            setattr(obj, attrname, attr)
+    if registry is not None:
+        for subclass in registry:
+            obj.register(subclass)
+
+    return obj
+
+
+def _reduce_global(pickler, obj):
+    """Custom reducing callback for functions and classes
+
+    This function is the analog of a custom save_global. However, the C Pickler
+    API does not expose low-level instructions such as save or write. Instead,
+    we return a reduce value the the Pickler will internally serialize via
+    save_reduce.
+    """
+    # XXX: This functions needs the current active pickler to pass its
+    # globals_ref to _function_reduce
+    t = type(obj)
+    try:
+        is_anyclass = issubclass(t, type)
+    except TypeError:  # t is not a class (old Boost; see SF #502085)
+        is_anyclass = False
+
+    if is_anyclass:
+        return _class_reduce(obj)
+    elif isinstance(obj, types.FunctionType):
+        return _function_reduce(obj, pickler.globals_ref)
```

Comment:
Because this is not a regular reducer either, maybe we could make `_function_reduce` a method of `CloudPickler` so that accessing `self.globals_ref` would be more natural.